teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,304 @@
+
+## Design Details
+
+`Tasks` can produce one or more execution `Results` under `/tekton/results/` by specifying `results`
+in the `Task` specification:
+
+```yaml
+spec:
+  results:
+    - name: leased-resource
+      description: The name of the leased resource
+  steps:
+```
+
+This task produces a result file at `/tekton/results/leased-resource`.
+
+Any other `Pipeline` `Tasks` can use these `Results` as parameter values through variable substitution:
+
+```yaml
+spec:
+  tasks:
+    - name: boskos-release
+      params:
+        - name: leased-resource
+          value: $(tasks.boskos-acquire.results.leased-resource)
+```
+
+No need to specify explicit runAfter here, `Tekton` infers the execution order by setting `boskos-acquire`
+as a dependency so that the task emitting the referenced results executes before the task that consumes them.
+
+Similar to non-final tasks, final task use `Results` through variable substitution:
+
+```yaml
+spec:
+ tasks:
+   - name: boskos-acquire
+     taskRef:
+       name: boskos-acquire
+   - name: do-stuff-with-resource
+     taskRef:
+       name: do-stuff-with-resource
+     params:
+       - name: resource-name
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+ finally:
+   - name: boskos-release
+     taskRef:
+       name: boskos-release
+     params:
+       - name: leased-resource
+         value:  $(tasks.boskos-acquire.results.leased-resource)
+```
+
+**Note:** Today final tasks are executed all in parallel and can not depend on any other `Task` in the `Pipeline`.
+With adding support for `Task` `Results`, we are introducing dependencies to finally `Tasks` but that
+does not change scheduling of finally tasks. All final tasks will still be executed in parallel after
+all non-final tasks finishes.
+
+**Q. What happens to `Pipeline` when the dependent task `boskos-acquire` either failed or not executed and
+the task result `leased-resource` is not initialized?**
+
+**A.** The finally `Task` `boskos-release` is attempted but declared failure since the task result
+`leased-resource` is not initialized. Pipeline controller logs the reason of this failure as invalid result
+reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with failure.
+
+**Q. What happens when the dependent `Task` succeeds but the `Task` `Result` is empty?**
+
+**A.** It could be possible that the task author intentionally chooses to leave the `Task` `Result` empty
+(creating an empty file at `/tekton/results/`). `Pipeline` leaves that param value empty and executes
+the final task. Empty `Task` `Result` is considered a valid value and results in successful variable
+substitution. The error handling of such empty `Task` `Result` is left to the final `Task` author.
+
+**Q. Can a final `Task` produce `Task` `Results`?**
+
+**A.** Nope, this proposal does not enable final tasks to produce `Task` `Results`.
+
+
+**Q. Explain the overall workflow.**
+
+**A.**
+
+```
```

Comment:
can you explain the context of this workflow a bit more? e.g. " Param Validation " i'm not sure if you mean, params to the finally tasks? params for all tasks?

(also im confused how we can validate the params before we apply results from dependent tasks?)

Reply:
In this context, "Param validation" does syntax check on params to all final tasks:

(1) check if parameter name is unique
(2) type of the default value matches with its type
(3) param with task results has valid result reference i.e. $(tasks.results.taskName.param)

"Validate and Apply results from Dependent Tasks" actually resolves the task result from dependent tasks and applies those values to params in the final tasks.

In general, the same workflow is executed for non-final tasks too.

Reply:
k i think this is a bit more clear to me